### DISCUSSION TITLE
save predict volume

### DISCUSSION THREAD

**Discussion by mikiwang820** [Project-MONAI/MONAI#1468] [Q&A] [answered]
Hi,

I wrote an evaluation to predict my labels for segmentation task and save as nifti file, but some outputs look like being shift.
Additionally, I got around 0.8 dice score for one patient before doing "saver.save_batch"  so that I think the predicational volume's position is correct at this moment.
After this, I saved it and calculated dice score with ground truth label and prediction, and now dice score turns to 0.57 because of image shift.
Is there any problem when I applied "saver.save_batch?
I also attach my evaluation code and how I calculate dice score.
I will appreciate your any help.

1) Evaluation code
```
crop_size = (96, 96, 32)
val_transforms = Compose(
    [
        LoadImaged(keys=["img", "seg"]),
        AddChanneld(keys=["img", "seg"]),
        Orientationd(keys=["img", "seg"], axcodes="RAS"),
        ScaleIntensityRanged(
            keys=["img"], a_min=-200, a_max=250, b_min=-1.0, b_max=1.0, clip=True,
        ),
        CropForegroundd(keys=["img", "seg"], source_key="img"),
        ToTensord(keys=["img", "seg"]),
    ]
)

val_ds = CacheDataset(data=val_files, transform=val_transforms, cache_rate=1.0, num_workers=4) 
val_loader = DataLoader(val_ds, batch_size=1, num_workers=4) 

dice_metric = DiceMetric(include_background=True, reduction="mean")
post_trans = Compose([Activations(sigmoid=True), AsDiscrete(threshold_values=True)])

device = torch.device("cuda:0")
model = UNet(
    dimensions=3,
    in_channels=1,
    out_channels=2,
    channels=(16, 32, 64, 128, 256),
    strides=(2, 2, 2, 2),
    num_res_units=2,
    norm=Norm.BATCH,
).to(device)

post_pred = Compose([AsDiscrete(argmax=True), KeepLargestConnectedComponent(applied_labels=[1], connectivity=1)])
output_dir = os.path.join(data_dir, "output")
model.load_state_dict(torch.load("/lits_1mm/monai_models/for_evaluation.pth"))
model.eval()

with torch.no_grad():
    metric_sum = 0.0
    metric_count = 0
    saver = NiftiSaver(output_dir)
    for val_data in val_loader:
        val_images, val_labels = val_data["img"].to(device), val_data["seg"].to(device)
        roi_size = crop_size
        sw_batch_size = 1
        #Evaluate
        val_outputs = sliding_window_inference(
            val_data["img"].to(device), roi_size, sw_batch_size, model
        )
        val_outputs = post_pred(val_outputs)
        #val_labels = post_label(val_labels)
        value, _ = dice_metric(y_pred=val_outputs, y=val_labels)
        metric_count += len(value)
        metric_sum += value.item() * len(value)
        saver.save_batch(val_outputs, val_data["img_meta_dict"])
    metric = metric_sum / metric_count
    print("evaluation metric:", metric)
```

2) My dice score code

```
import os
import glob
import SimpleITK as sitk
import numpy as np

dice_sum = 0

def binary_dice3d(s,g):
    #dice score of two 3D volumes
    num=np.sum(np.multiply(s, g))
    denom=s.sum() + g.sum() 
    if denom==0:
        return 1
    else:
        return  2.0*num/denom
    
GT_route = "/Users/wangmeiqi/code_and_data/HCC/temp/true"
predict_route = "/Users/wangmeiqi/code_and_data/HCC/temp/predict"
GT_images = sorted(glob.glob(os.path.join(GT_route, "*.nii.gz")))
predict_images = sorted(glob.glob(os.path.join(predict_route, "*.nii.gz")))
save_route = "/Users/wangmeiqi/code_and_data/HCC/temp"


for i in range(len(GT_images)):
    
    read_gt = sitk.ReadImage(GT_images[i])
    GT_array = sitk.GetArrayFromImage(read_gt)
    predict_array = sitk.GetArrayFromImage(sitk.ReadImage(predict_images[i]))
    dice = binary_dice3d(predict_array, GT_array)
    print(predict_images[i].split("/")[-1] + ": " + str(dice))
    dice_sum = dice_sum + dice

avg_dice = dice_sum/len(GT_images)

```
3) Output volume
<img width="357" alt="截圖 2021-01-19 下午10 09 00" src="https://user-images.githubusercontent.com/61928785/105045714-2f9e3a80-5aa3-11eb-9b65-0c4d7674ccd5.png">

**Comment by rijobro** [comment, marked as answer]
Am I right in thinking that you crop your segmentations to a size of 96x96x32, yet (I'm guessing here) you superpose that on top of your original, non-cropped image? If so, this will be where the difference is coming in.

**Reply by mikiwang820** [reply]
Actually, I only use crop_size = (96, 96, 32) in "sliding_window_inference" function as roi_size. I think maybe that is different from cropping my volumes.

**Reply by rijobro** [reply]
Ok, well the first thing to check is the image metadata. It looks like you're viewing the images in ITKSnap? You can load up the image info (I think it's Ctrl+I), and you'll want to check image sizes, affine transformations and offsets.

If you run your coded in something like jupyter notebook or spyder, you could also view the images directly to see if the segmentations nicely match up with the input images there. That would help narrow down whether the offset is coming from the image transformations or the Nifti saver.

**Reply by mikiwang820** [reply]
To be honest, my image's metadata (Fig. 1) is not completely the same with ground truth label (Fig. 2) and predictive label (Fig. 3) , but between these two are the same. They still cannot match up (Fig. 4), that's the reason why I am really confused to this problem. Actually, I have tried to use SimpleITK to copy images' header to predictive labels', while that doesn't work.
And I also viewed the images in Spyder (Fig. 4, 5).

1) image's metadata
<img width="634" alt="image" src="https://user-images.githubusercontent.com/61928785/105282977-764a7c80-5bea-11eb-909a-cf899b8a27f4.png">

2) ground truth label's metadata
<img width="634" alt="GT" src="https://user-images.githubusercontent.com/61928785/105283022-8f532d80-5bea-11eb-867f-017e1654eb63.png">

3) predictive label's metadata
<img width="634" alt="prediction" src="https://user-images.githubusercontent.com/61928785/105283040-9d08b300-5bea-11eb-9ff4-5c932387ef31.png">

4) GT and prediction viewed in Spyder
<img width="304" alt="截圖 2021-01-21 下午1 05 20" src="https://user-images.githubusercontent.com/61928785/105283074-ad209280-5bea-11eb-908f-4934bb4801cf.png">

5) image and prediction viewed in Spyder
<img width="304" alt="截圖 2021-01-21 下午1 05 14" src="https://user-images.githubusercontent.com/61928785/105283118-c590ad00-5bea-11eb-9028-cbd5e4bbdfdd.png">

**Reply by rijobro** [reply]
So in fig. 5, you show the image and the predicted segmentation, without either having been saved to disk? i.e., the image went through the DataLoader (meaning it had transforms applied to it), then you ran the model to get the predicted segmentation, ran the post_transform and then displayed the two images? 

If that is the case, then the problem isn't obvious to me. One idea might be to start with [the 3D spleen segmentation notebook](https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d.ipynb), since the labels seem to nicely superpose, and then gradually modify it to your needs?

**Reply by mikiwang820** [reply]
Sorry, I think I misunderstood before. However, I check my images and labels again before doing post transfer. As shown in the below figure, the image and label are matched. Therefore, I guess the problem maybe comes from the post transform or I didn't use post transform functions in a right way?

1) the brighter one is predicted label for different patients
![image](https://user-images.githubusercontent.com/61928785/105355657-64d69400-5c2d-11eb-9f15-ead9ae1e3513.png)

**Reply by rijobro** [reply]
Hi, yes in that case it looks like the discrepancy is due to writing the segmentations at the end. Possibly related to this issue in that case: https://github.com/Project-MONAI/MONAI/issues/365.

Earlier you sent a few screenshots from ITKSnap, but it's missing pixdim[1] and pixdim[2] (i.e., voxel spacing in x- and y-directions), and also the number of voxels in all directions. Would you be able to send that information?

**Reply by mikiwang820** [reply]
Thank you so much @rijobro 
Here is the information for pixdim[1] and pixdim[2]. 
1) image
<img width="637" alt="image" src="https://user-images.githubusercontent.com/61928785/105439658-3776fe00-5ca0-11eb-8e91-d8d4f9745050.png">

2) ground truth label
<img width="637" alt="gt" src="https://user-images.githubusercontent.com/61928785/105439689-4362c000-5ca0-11eb-903e-4ea6d70104d1.png">

3) predicted label
<img width="637" alt="predict" src="https://user-images.githubusercontent.com/61928785/105439713-4d84be80-5ca0-11eb-84f6-6b5fbe88118e.png">

**Reply by mikiwang820** [reply]
Hi @rijobro ,
I finally find my problem.
I use "monai.data.Dataset" instead of "CacheDataset", it works.
Thanks for your help.

**Reply by rijobro** [reply]
I'm surprised that solves the problem, but glad that it does!
